bookkeeper-server/src/main/java/org/apache/bookkeeper/client/LedgerHandle.java

```diff
@@ -1898,302 +1788,81 @@ void handleBookieFailure(final Map<Integer, BookieSocketAddress> failedBookies)
-                        getId(), ensembleChangeIdx);
-                }
-
-                rereadMetadata(new ReReadLedgerMetadataCb(rc,
-                                       ensembleInfo, curBlockAddCompletions, ensembleChangeIdx));
-                return;
-            } else if (rc != BKException.Code.OK) {
-                LOG.error("[EnsembleChange-L{}-{}] : could not persist ledger metadata : info = {}, "
-                        + "closing ledger : {}.", getId(), ensembleChangeIdx, ensembleInfo, rc);
-                handleUnrecoverableErrorDuringAdd(rc);
-                return;
-            }
-            int newBlockAddCompletions = blockAddCompletions.decrementAndGet();
 
-
-            if (LOG.isDebugEnabled()) {
-                LOG.info("[EnsembleChange-L{}-{}] : completed ensemble change, block add completion {} => {}",
-                        getId(), ensembleChangeIdx, curBlockAddCompletions, newBlockAddCompletions);
-            }
-
-            // We've successfully changed an ensemble
-            ensembleChangeCounter.inc();
-            LOG.info("New Ensemble: {} for ledger: {}", ensembleInfo.newEnsemble, ledgerId);
-
-            // the failed bookie has been replaced
-            unsetSuccessAndSendWriteRequest(ensembleInfo.newEnsemble, ensembleInfo.replacedBookies);
-        }
-
-        @Override
-        public String toString() {
-            return String.format("ChangeEnsemble(%d)", ledgerId);
-        }
-    }
-
-    /**
-     * Callback which is reading the ledgerMetadata present in zk. This will try
-     * to resolve the version conflicts.
-     */
-    private final class ReReadLedgerMetadataCb extends OrderedGenericCallback<LedgerMetadata> {
-        private final int rc;
-        private final EnsembleInfo ensembleInfo;
-        private final int curBlockAddCompletions;
-        private final int ensembleChangeIdx;
-
-        ReReadLedgerMetadataCb(int rc,
-                               EnsembleInfo ensembleInfo,
-                               int curBlockAddCompletions,
-                               int ensembleChangeIdx) {
-            super(clientCtx.getMainWorkerPool(), ledgerId);
-            this.rc = rc;
-            this.ensembleInfo = ensembleInfo;
-            this.curBlockAddCompletions = curBlockAddCompletions;
-            this.ensembleChangeIdx = ensembleChangeIdx;
-        }
-
-        @Override
-        public void safeOperationComplete(int newrc, LedgerMetadata newMeta) {
-            if (newrc != BKException.Code.OK) {
-                LOG.error("[EnsembleChange-L{}-{}] : error re-reading metadata "
-                                + "to address ensemble change conflicts: {}",
-                        ledgerId, ensembleChangeIdx, BKException.codeLogger(newrc));
-                handleUnrecoverableErrorDuringAdd(rc);
+        synchronized (metadataLock) {
+            if (changingEnsemble) {
+                delayedWriteFailedBookies.putAll(failedBookies);
```

Comment:
I could be misunderstanding, but I think the main difference in behavior is that between when changingEnsemble gets set to true and when the ensemble change completes, new writes continue to use the old ensemble whereas with the old machinery they would have optimistically used the new one.  In both cases, we have to delay acks until the ensemble change is complete, but with this variant we may have to wait to resend writes for entries written since the ensemble change began which didn't get aQ responses from unchanged indexes.  Am I understanding that correctly @ivankelly ?

Reply:
@jvrao Something to note here is that unsetSuccessAndSendWriteRequest will resend the write request to those bookies regardless of whether the entry already has aQ copies, so it shouldn't generate additional under replicated ledgers in the common case.

Reply:
I'm not really worried about that, and we can always add machinery for projecting the post-update ensemble for writes later.

Reply:
>  We could modify this to unsetSuccessAndSendWriteRequest after each sucessful zk write.

Aren't we doing that now? If there is a change ensemble in flight we are putting failed bookies back to delayedWriteFailedBookies(), and leaving from here. So when the current ensemble change comes back we call unsetSuccessAndSendWriteRequest(). But the problem is we don't attempt to change ensemble with delayedWriteFailedBookies() until next write comes in. What if the next write will take long time to come in? We may fire writeQuorumTimeout without even attempting an ensemble change.

Reply:
Ah I see you are doing the loop in whenCompleteAsync(); Never mind; premature comment. Will delete it.

Reply:
make senses to me